pkg/storage/gcp/saver.go

```diff
@@ -20,47 +18,16 @@ import (
 // Uploaded files are publicly accessable in the storage bucket as per
 // an ACL rule.
 func (s *Storage) Save(ctx context.Context, module, version string, mod []byte, zip io.Reader, info []byte) error {
-	if exists := s.bucket.Exists(ctx, module, version); exists {
+	if exists := s.Exists(module, version); exists {
 		return stg.ErrVersionAlreadyExists{Module: module, Version: version}
 	}
 
-	errs := make(chan error, 3)
-	// create a context that will time out after the value found in
-	// the ATHENS_TIMEOUT env variable
-	ctxWT, cancelCTX := context.WithTimeout(ctx, env.Timeout())
-	defer cancelCTX()
-
-	// dispatch go routine for each file to upload
-	go upload(ctxWT, errs, s.bucket, module, version, "mod", "text/plain", bytes.NewReader(mod))
-	go upload(ctxWT, errs, s.bucket, module, version, "zip", "application/octet-stream", zip)
-	go upload(ctxWT, errs, s.bucket, module, version, "info", "application/json", bytes.NewReader(info))
-
-	var errors error
-	// wait for each routine above to send a value
-	for count := 0; count < 3; count++ {
-		err := <-errs
-		if err != nil {
-			errors = multierror.Append(errors, err)
-		}
-	}
-	close(errs)
-
-	return errors
-}
-
-// upload waits for either writeToBucket to complete or the context expires
-func upload(ctx context.Context, errs chan<- error, bkt Bucket, module, version, ext, contentType string, file io.Reader) {
-	select {
-	case errs <- writeToBucket(ctx, bkt, module, version, ext, contentType, file):
-		return
-	case <-ctx.Done():
-		errs <- fmt.Errorf("WARNING: context deadline exceeded during write of %s version %s", module, version)
-	}
+	err := modupl.Upload(ctx, module, version, bytes.NewReader(info), bytes.NewReader(mod), zip, s.upload)
```

Comment:
maybe you can just return the function call here

Reply:
We need to update the list file before we return so I've put a comment here.